File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.scala
// import scala.math.min
// object Solution {
//     def minCost(colors: String, neededTime: Array[Int]): Int = {
//         val colIdx= colors.zipWithIndex
//         colIdx.sliding(2).map{
//             case Seq((firstColor, firstIndex),(secondColor, secondIndex))=>{
//                 if(firstColor== secondColor) min(neededTime(firstIndex), neededTime(secondIndex))
//                 else 0
//             }
//         }.toSeq.sum
        
//     }
// }

object Solution {
    def solve(sum: Int, prev: Int, i: Int, colors: String, neededTime: Array[Int]): Int = {
        if (i < colors.length) {
            if (colors(prev) != colors(i))
                solve(sum, i,i + 1, colors, neededTime)
            else {
                val newSum = sum + math.min(neededTime(prev), neededTime(i))
                if(math.min(neededTime(prev), neededTime(i))== neededTime(i)){
                    solve(newSum, prev, i + 1, colors, neededTime)
                }
                else    
                    solve(newSum, i, i + 1, colors, neededTime)
            }
        }
        else {
          sum
        }
    }
    def minCost(colors: String, neededTime: Array[Int]): Int = {
        val res:Int=solve(0,0,1,colors,neededTime)
        res
    }
}

File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.java
class Solution {
    public int minCost(String colors, int[] nt) {
        int ans=0;
        int n= colors.length();
        char pr= colors.charAt(0);
        int pri=0;
        for(int i=1; i< n; i++){
            if(colors.charAt(i)== pr){
                int min= Math.min(nt[i], nt[pri]);
                if(min== nt[pri]){
                    pri= i;
                }
                ans+= min;
            }
            else{
                pr= colors.charAt(i);
                pri= i;
            }
        }
        return ans;
    }
}